jmespath-core/src/main/java/io/burt/jmespath/function/CompareByFunction.java

```diff
@@ -31,14 +31,23 @@ public CompareByFunction() {
     if (elements.hasNext()) {
       T result = elements.next();
       T resultValue = expression.search(result);
-      boolean expectNumbers = expectNumbers(runtime, resultValue);
+      JmesPathType resultValueType = runtime.typeOf(resultValue);
+      boolean expectNumbers = true;
+      if (resultValueType == JmesPathType.STRING) {
+        expectNumbers = false;
+      } else if (resultValueType != JmesPathType.NUMBER) {
+        return runtime.handleArgumentTypeError(this, "number or string", resultValueType.toString());
+      }
       while (elements.hasNext()) {
         T candidate = elements.next();
         T candidateValue = expression.search(candidate);
-        checkType(runtime, candidateValue, expectNumbers);
-        if (sortsBefore(runtime.compare(candidateValue, resultValue))) {
-          result = candidate;
-          resultValue = candidateValue;
+        if (checkType(runtime, candidateValue, expectNumbers)) {
+          if (sortsBefore(runtime.compare(candidateValue, resultValue))) {
+            result = candidate;
+            resultValue = candidateValue;
+          }
+        } else {
+          return runtime.handleArgumentTypeError(this, expectNumbers ? "number" : "string", runtime.typeOf(candidateValue).toString());
```

Comment:
This method is very similar to that in [SortByFunction.java](https://github.com/burtcorp/jmespath-java/blob/avoid-function-call-exceptions/jmespath-core/src/main/java/io/burt/jmespath/function/SortByFunction.java), except from a few variable names and the fact that the latter also builds a list of all pairs of raw and transformed array elements. I tried extracting the common functionality myself, but gave up when I realized it wasn't as trivial as I'd initially thought. I still believe it'd be possible to do it though, with a little bit of thought.